messagingpush/src/main/java/io/customer/messagingpush/ModuleMessagingPushFCM.kt

```diff
@@ -0,0 +1,37 @@
+package io.customer.messagingpush
+
+import io.customer.messagingpush.di.fcmTokenProvider
+import io.customer.sdk.CustomerIOInstance
+import io.customer.sdk.CustomerIOModule
+import io.customer.sdk.di.CustomerIOComponent
+
+class ModuleMessagingPushFCM : CustomerIOModule {
```

Comment:
I don't think it needs to be this complicated, we can easily avoid storing refrences of `CustomerIOInstance` `CustomerIOComponent` and relying on them. 

I agree with the approach but I would recommend something along this line.

Suggestion, 

`CustomerIOConfig` takes in `DeviceTokenGenerator` object in its builder. 

```
interface DeviceGenerator {
    public fun asyncGenerateDeviceToken(onDeviceTokenGenerated: (token: String) -> Unit)
}
```

Messaging module has FCM implementation of the `DeviceGenerator`. 

And then the client can simply add our `messaging module` and in builder provide its device generator implementation. 

so instead of `addCustomerIOModule` we hae `addCustomerDeviceGenerator`. 

Let me know what you think, I would like to keep both modules agnostic and not hold the reference.

Reply:
I do agree that the implementation in this PR is overkill for the goal of simply getting the push token each time the SDK initializes. 

The reason I made this implementation this way was to begin introducing the idea of "module hooks" like we have on the iOS SDK today. "module hooks" meaning a way for the CIO SDKs (tracking, messaging-push-fcm) to communicate together at runtime. Loosely couple them together to let them communicate but keep the various SDKs remain optional to install and not cause circular dependencies together. 

This PR is proposing that customers make 1 change to their app to enable this feature:
```kotlin
CustomerIO.Builder()
.addCustomerIOModule(ModuleMessagingPushFCM())
.build()
```

That is more future-proof in that we can now add more features to the FCM messaging push module without asking customers to update their app's code. 

If we were to instead ask customers to enable this 1 feature in their code like this:
```kotlin
CustomerIO.Builder()
.asyncGenerateDeviceToken(CustomerIOFCMDeviceTokenGenerator())
.build()
```

Then when we add a new feature to the SDK in the messaging push FCM module, we may need to ask customers to add more code to their apps (because we would ask them to add more code to the `CustomerIO.Builder()`). 

I am imagining the SDK's long-term vision enabling and configuring a series of modules. Something like this:
```kotlin
CustomerIO.Builder()
.addCustomerIOModule(ModuleMessagingPushFCM(
    // you can configure each of the CIO SDKs like this:
    getCurrentTokenOnInit = true
    // or with a .Builder() to be Java compatible. 
))
// As we add more optional SDKs for customers to install, they just enable these SDKs this way:
.addCustomerIOModule(ModuleInAppPush())
.build()
```

This PR is the foundation of initializing SDK modules and enabling "module hooks" between the various modules.

Reply:
I think this might be a broader conversation but `hooks` made more sense when we could have multiple instances, but now that SDK is a singleton and hopefully uni-flow, as in the `base` module don't need to know what `feature` modules are doing and won't be passing messages to it.

It's only feature modules that will need to call methods of `base/client` artifact which they can use 
`CustomerIO.instance()`.

I want to really understand the use-case if we have any or are in the pipeline to hold the references of the whole module as they contain further references of all the members as well and are heavy.